### PR TITLE
Port free_block()

### PR DESCRIPTION
**Pull request by XrXr**
Should be good assuming CI passes. Might be easier to go commit by commit.

### CODE REVIEW

**Comment by XrXr** [comment]
For [3090776](https://github.com/Shopify/ruby/pull/240/commits/30907760b66a824614667a000a6a8408caf3edf6), it's probably fairly easy to refactor and pass around a `&mut Invariant` so we can get it statically borrow checked. Probably a separate PR though.

**Comment by maximecb** [comment]
> For [3090776](https://github.com/Shopify/ruby/pull/240/commits/30907760b66a824614667a000a6a8408caf3edf6), it's probably fairly easy to refactor and pass around a `&mut Invariant` so we can get it statically checked. Probably a separate PR though.

We may want to implement the logic for `rb_yjit_bop_redefined` as a method of `Invariant` 🤔  ?

File: yjit/src/core.rs [hunk 1]
```diff
@@ -1776,55 +1789,42 @@ pub fn defer_compilation(jit: &JITState, cur_ctx: &Context, cb: &mut CodeBlock,
...
             }
         }
-
-        // Free the outgoing branch entry
-        free(out_branch);
     }
 
-    rb_darray_free(block->incoming);
-    rb_darray_free(block->outgoing);
-    rb_darray_free(block->gc_object_offsets);
```
Comment (maximecb) [inline yjit/src/core.rs:1824]: Should we be clearing these arrays here (setting their length to zero)? Just to drop the references and indicate that this block has been freed.

Comment (XrXr) [reply yjit/src/core.rs:1824]: Hmm in the event that we have lingering references to the block which prevents deallocation having these info might be useful for debugging. Maybe we should add a debug only invalidated field though. These are automatically dropped when the block is dropped.

**Pull request merged**